### PR TITLE
ADEN-2937 Prebid.js integration

### PR DESCRIPTION
**Pull request by fraszczakszymon**

### CODE REVIEW

File: extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+				}
+			]
+		};
+	}
+
+	function getAdUnits(skin) {
+		var adUnits = [],
+			slotName;
+
+		if (!slots[skin]) {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js]: `if`statement is redundant -- in any case function returns `[]`

File: extensions/wikia/AdEngine/js/lookup/adapter/matomy.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			]
+		};
+	}
+
+	function configureOasisSlots() {
+		var context = adContext.getContext(),
+			i,
+			slotName;
+
+		if (context.targeting.pageType === 'home') {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/adapter/matomy.js]: It's the same as https://github.com/Wikia/app/blob/dev/extensions/wikia/AdEngine/js/lookup/rubiconFastlane.js#L129.
Do you have any ideas how we can generalize it?

File: extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			return [];
+		}
+
+		for (slotName in slots[skin]) {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js]: Can we use `forEach`?
Like here: https://github.com/Wikia/app/pull/10240/files#diff-7a14ff9954fe20151dc1f68b216f2d20R77

File: extensions/wikia/AdEngine/js/lookup/prebid.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		sizeKey = 'hb_size',
+		url = '//acdn.adnxs.com/prebid/prebid.js';
+
+	function addAdUnits(adapterAdUnits) {
+		adapterAdUnits.forEach(function (adUnit) {
+			adUnits.push(adUnit);
+		});
+	}
+
+	function setupAdUnits(skin) {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:26]: What is the point of moving it to separate function if you call it once?

File: extensions/wikia/AdEngine/js/lookup/prebid.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+				win.pbjs.addAdUnits(adUnits);
+
+				win.pbjs.requestBids({
+					bidsBackHandler: onResponse
+				});
+			});
+		}
+	}
+
+	function encodeParamsForTracking(params) {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:60]: I would invert it:

```
    function encodeParamsForTracking(params) {
        if (!params[bidderKey] || !params[sizeKey] || !params[bidKey]) {
            return '';
        }

        return [params[bidderKey], params[sizeKey], params[bidKey]].join(';');
    }
```

File: extensions/wikia/AdEngine/js/lookup/prebid.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+], function (appnexus, matomy, factory, doc, win) {
+	'use strict';
+
+	var adapters = [
+			appnexus,
+			matomy
+		],
+		adUnits = [],
+		priceMap = {},
+		bidderKey = 'hb_bidder',
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:15]: What is the difference between `bidderKey` and `bidKey`; why do we need both?

File: extensions/wikia/AdEngine/js/lookup/prebid.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		return priceMap;
+	}
+
+	function isSlotSupported(slotName) {
+		return adUnits.some(function (adUnit) {
+			return adUnit.code === slotName;
+		});
+	}
+
+	function getSlotParams(slotName) {
```
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:93]: What about:

```
    function getSlotParams(slotName) {
        var params;

        if (win.pbjs && typeof win.pbjs.getAdserverTargetingForAdUnitCode === 'function') {
            params = win.pbjs.getAdserverTargetingForAdUnitCode(slotName);
        }

        return params || {};
    }
```

?

File: extensions/wikia/AnalyticsEngine/AnalyticsProviderPrebid.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+class AnalyticsProviderPrebid implements iAnalyticsProvider {
```
Comment (drets) [inline extensions/wikia/AnalyticsEngine/AnalyticsProviderPrebid.php:3]: 👍

**Comment by drets** [comment]
Minor comments, otherwise LGTM 👍

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 6 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:15]: bid != bidder
bidder - param with adapter name
bid - param with price name

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 4 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:26]: To make it more readable

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 7 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:93]: 👍

File: extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js (hunk 3 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js]: 👍

File: extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js (hunk 1 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/adapter/appnexus.js]: :+1:

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 5 above)
Comment (fraszczakszymon) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:60]: Ok 👍

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 6 above)
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:15]: Thanks for explanation 😉

File: extensions/wikia/AdEngine/js/lookup/prebid.js (hunk 4 above)
Comment (drets) [inline extensions/wikia/AdEngine/js/lookup/prebid.js:26]: 👍

File: tests/karma/js-unit.conf.js [hunk 9]
```diff
@@ -89,6 +89,7 @@ module.exports = function (config) {
 			'extensions/wikia/AdEngine/js/config/*.js',
 			'extensions/wikia/AdEngine/js/context/*.js',
 			'extensions/wikia/AdEngine/js/lookup/*.js',
```
Comment (fraszczakszymon) [inline tests/karma/js-unit.conf.js]: Is this line still needed?

**Comment by fraszczakszymon** [comment]
LGTM 👍

File: tests/karma/js-unit.conf.js (hunk 9 above)
Comment (bognix) [inline tests/karma/js-unit.conf.js]: no, removed

**Comment by fraszczakszymon** [comment]
LGTM 👍

**Pull request merged**